### PR TITLE
SQL-991: Add SQLDriverConnect edge case handling

### PR DESCRIPTION
**Pull request by bucaojit**
Added error handling in SQLDriverConnect for:
- 01004: Output string being right truncated due to buffer not being large enough
- 01S00: Invalid connection string causing Rust driver to fail on parse
- HYC00: Driver completion options not being supported: SQL_DRIVER_PROMPT, SQL_DRIVER_COMPLETE or SQL_DRIVER_COMPLETE_REQUIRED
- HYT00: Timeout expired
- IM007: No Driver property specified

Added unit tests for the errors that occur before the actual connect call.  The errors that occur after/during the connect are `01004` and `HYT00`, which may need to be done at integration test time to have the ADF running to connect to.

I did not find edge cases to add for the SQLDisconnect function.  One that looked relevant is `08003 | Connection not open`, however there wasn't a call on the `client` object to check whether the connection was open.

### CODE REVIEW

File: core/src/conn.rs [hunk 1]
```diff
@@ -42,7 +43,7 @@ impl MongoConnection {
         login_timeout: Option<u32>,
         application_name: Option<&str>,
     ) -> Result<Self> {
-        let mut client_options = ClientOptions::parse(mongo_uri)?;
+        let mut client_options = ClientOptions::parse(mongo_uri).map_err(Error::MongoParseError)?;
```
Comment (pmeredit) [inline core/src/conn.rs:46]: nice!

File: core/src/conn.rs [hunk 2]
```diff
@@ -9,8 +10,8 @@ pub struct MongoConnection {
     pub client: Client,
     /// The current database set for this client.
     /// All new queries will be done on this DB.
-    /// We stick with mongo terminalogy here and ODBC
-    /// terminalogy in the odbc wrappers, hence
```
Comment (pmeredit) [inline core/src/conn.rs:13]: Seems I need to get spell check running in nvim

File: core/src/err.rs [hunk 3]
```diff
@@ -1,30 +1,35 @@
...
 impl Error {
     pub fn get_sql_state(&self) -> &'static str {
         match self {
-            // TODO: for now we just return HY000 for all Mongo Errors.
-            // In the future this will change based on the type of error.
-            // This should be updated as we introduce features that can refine
-            // this.
-            Error::MongoError(_) => HY000,
+            Error::MongoError(err) => {
+                if matches!(err.kind.as_ref(), ErrorKind::Io(ref io_err) if io_err.kind() == std::io::ErrorKind::TimedOut)
```
Comment (pmeredit) [inline core/src/err.rs:19]: Huh, I didn't know `matches!` had a conditional. Very nice.

**Review by pmeredit — APPROVED** [review]
LGTM!

File: odbc/src/api/driver_connect_tests.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+
+        let in_connection_string =
+            "Driver=ADF_ODBC_DRIVER;USER=N_A;SERVER=N_A;AUTH_SRC=N_A;PWD=N_A";
+        // Parse error due to illegal character in SERVER field
+        driver_connect_check_diagnostics(
+            "PWD=N_A;Driver=ADF_ODBC_DRIVER;UID=N_A;SERVER=//;AUTH_SRC=N_A;USER=N_A",
+            DriverConnectOption::NoPrompt,
+            _01S00,
```
Comment (nbagnard) [inline odbc/src/api/driver_connect_tests.rs:78]: Sorry, this was a mistake on my part. I missed the fact that for 01S00 the driver is able to connect to the data source anyway and the function returns SQL_SUCCESS_WITH_INFO.
We can use 08001 (Client unable to establish connection) or HY000 (General error).
Feel free to go over the list of errors and description and if you find one which is more fitting, let me know.

**Review by nbagnard — CHANGES_REQUESTED** [review]
This is looking good so far.
I have a couple of comments, but certainly nothing major.

File: odbc/src/api/functions.rs [hunk 5]
```diff
@@ -450,6 +451,9 @@ fn sql_driver_connect(
     let mut odbc_uri = ODBCUri::new(odbc_uri_string)?;
     let mongo_uri = odbc_uri.remove_to_mongo_uri()?;
     let auth_src = odbc_uri.remove_or_else(|| "admin", &["auth_src"]);
+    odbc_uri
+        .remove(&["driver"])
+        .ok_or(ODBCError::MissingDriverProperty)?;
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply odbc/src/api/functions.rs:456]: Oh, thanks for catching that.  I took this line in the design too literally and thought it meant we only supported `driver`: [Only support connection string with DRIVER=xxx](https://docs.google.com/document/d/1BiJZm6dZVYNlnkNh8nPHGHWOoXLWkzmiCuYtGKyqzu8/edit#bookmark=id.6mx2ey10b3hm).  I missed the next box that shows `Support for DSN`,  I'll add the dsn here.

File: odbc/src/api/driver_connect_tests.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            "Driver=ADF_ODBC_DRIVER;SERVER=N_A;AUTH_SRC=N_A;PWD=N_A",
+            DriverConnectOption::NoPrompt,
+            HY024,
+            SqlReturn::ERROR,
+            "[MongoDB][API] Invalid Uri One of [\"user\", \"uid\"] is required for a valid Mongo ODBC Uri",
+        );
+
+        // Missing 'PWD'
+        driver_connect_check_diagnostics(
+            "Driver=ADF_ODBC_DRIVER;UID=N_A;SERVER=N_A;AUTH_SRC=N_A;USER=N_A",
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs:95]: Good catch, this is a copy/paste bug on my part.  The test wasn't meant to have both UID/USER but I have it in a few of the connection strings, I'll remove them. 🐛

File: odbc/src/api/driver_connect_tests.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                in_connection_string_encoded.as_ptr(),
+                in_connection_string.len().try_into().unwrap(),
+                out_connection_string,
+                buffer_length,
+                string_length_2,
+                driver_completion,
+            );
+            assert_eq!(expected_sql_return, actual_return_val);
+
+            let text_length_ptr = &mut 0;
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs:45]: Created [SQL-1036](https://jira.mongodb.org/browse/SQL-1036) to address this.

File: odbc/src/api/driver_connect_tests.rs (hunk 4 above)
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs]: I changed this to `08001`,  UNABLE_TO_CONNECT

File: odbc/src/api/driver_connect_tests.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            SqlReturn::ERROR,
+            "[MongoDB][Core] Parse error An invalid argument was provided: illegal character in database name",
+        );
+
+        // Missing Parameters in connection string
+        // Missing 'USER'
+        driver_connect_check_diagnostics(
+            "Driver=ADF_ODBC_DRIVER;SERVER=N_A;AUTH_SRC=N_A;PWD=N_A",
+            DriverConnectOption::NoPrompt,
+            HY024,
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs]: Sounds good.  I changed this to 08001, UNABLE_TO_CONNECT.

File: odbc/src/api/driver_connect_tests.rs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                &mut 0,
+                actual_message_text,
+                512,
+                text_length_ptr,
+            );
+            let actual_message_length = *text_length_ptr as usize;
+            unsafe {
+                let output = String::from_utf16_lossy(&*(actual_message_text as *const [u16; 256]));
+                let substring = &output[0..actual_message_length];
+                assert_eq!(expected_error_message, substring);
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs]: Done.

File: core/src/err.rs [hunk 10]
```diff
@@ -1,30 +1,35 @@
...
+use constants::{HY000, HYT00, _01S00};
 use mongodb::error::{BulkWriteFailure, ErrorKind, WriteFailure};
 use thiserror::Error;
 
 pub type Result<T> = std::result::Result<T, Error>;
 
 #[derive(Error, Debug)]
 pub enum Error {
+    #[error("Parse error {0}")]
+    MongoParseError(mongodb::error::Error),
```
[the remark this answers is not in the document]
Comment (bucaojit) [reply core/src/err.rs]: That works.  I changed it to: `Invalid connection string. Parse error: {0}` `MongoParseConnectionStringError`

File: constants/src/lib.rs [hunk 11]
```diff
@@ -1,7 +1,11 @@
...
 
 // SQL states
-pub const HYC00: &str = "HYC00";
-pub const HY000: &str = "HY000";
-pub const HY024: &str = "HY024";
-pub const _01S02: &str = "01S02";
+pub const NOT_IMPLEMENTED: &str = "HYC00";
+pub const TIMEOUT_EXPIRED: &str = "HYT00";
+pub const GENERAL_ERROR: &str = "HY000";
+pub const INVALID_VALUE: &str = "HY024";
```
Comment (nbagnard) [inline constants/src/lib.rs:7]: [nit] What about naming it INVALID_ATTR_VALUE instead of INVALID_VALUE so that it is clear that this is related to a value from a result set?

File: odbc/src/api/errors.rs [hunk 12]
```diff
@@ -1,16 +1,34 @@
...
 #[derive(Debug, Error)]
 pub enum ODBCError {
     #[error("[{}][API] The feature {0} is not implemented", VENDOR_IDENTIFIER)]
     Unimplemented(&'static str),
+    #[error(
+        "[{}][API] The driver connect option {0} is not supported",
+        VENDOR_IDENTIFIER
+    )]
+    UnsupportedDriverConnectOption(String),
     #[error("[{}][API] Invalid Uri {0}", VENDOR_IDENTIFIER)]
```
Comment (nbagnard) [inline odbc/src/api/errors.rs:16]:
```suggestion
    #[error("[{}][API] Invalid Uri: {0}", VENDOR_IDENTIFIER)]
```

File: odbc/src/api/driver_connect_tests.rs [hunk 13]
```diff
@@ -0,0 +1,150 @@
+use crate::{handles::definitions::*, SQLDriverConnectW, SQLGetDiagRecW};
+use constants::{NOT_IMPLEMENTED, NO_DSN_OR_DRIVER, UNABLE_TO_CONNECT};
+use odbc_sys::{DriverConnectOption, HandleType, SqlReturn};
+use std::sync::RwLock;
+
+mod unit {
+    use super::*;
+    use odbc_sys::SmallInt;
+    #[test]
+    fn connect_edge_cases() {
```
Comment (nbagnard) [inline odbc/src/api/driver_connect_tests.rs:10]: It would probably be better to extract each test so that the tests don't stop on the first failed assertion.

**Review by nbagnard — CHANGES_REQUESTED** [review]

File: constants/src/lib.rs (hunk 11 above)
Comment (bucaojit) [reply constants/src/lib.rs]: Sounds good, made the naming change to `INVALID_ATTR_VALUE`

File: odbc/src/api/driver_connect_tests.rs (hunk 13 above)
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs]: That's a good idea.  I changed the function `driver_connect_check_diagnostics()` into a macro and now each of the tests here are run separately.

Comment (nbagnard) [reply odbc/src/api/driver_connect_tests.rs]: Perfect. Thank you :)

File: odbc/src/api/driver_connect_tests.rs [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+                    String::from_utf16(&*(expected_sql_state_encoded.as_ptr() as *const [u16; 6]))
+                        .unwrap(),
+                    String::from_utf16(&*(actual_sql_state as *const [u16; 6])).unwrap()
+                );
+            }
+        }
+    };
+}
+
+mod unit {
```
Comment (nbagnard) [inline odbc/src/api/driver_connect_tests.rs:76]: I am actually not sure that these can really be called unit tests, it sounds more like integration tests. Maybe they should move to /odbc/tests/connection_tests.rs?
And the part that checks the diagnostics could go into the odbc/tests/common/mod.rs.
Unit tests could be added, if you think it's necessary, targeting sql_driver_connect and simply checking if the result is OK or Error.
Looks like this is something for [SQL-1036](https://jira.mongodb.org/browse/SQL-1036).
WDYT?

**Review by nbagnard — APPROVED** [review]

File: odbc/src/api/driver_connect_tests.rs (hunk 14 above)
Comment (bucaojit) [reply odbc/src/api/driver_connect_tests.rs:76]: That sounds reasonable to me.  
I'll add these steps to SQL-1036 and we can update the testing at that point.

**Comment by bucaojit** [comment]
evergreen merge
